editoast/src/views/timetable/track_occupancy_utils.rs

```diff
@@ -0,0 +1,269 @@
+            }
+        }
+    }
+    None
+}
+
+/// Extract track ranges and path item positions from pathfinding result
+/// Returns None if pathfinding failed
+///
+/// Note: The timetable can have a pathfinding failure for a train schedule
+/// because the train schedule is not valid. We skip it because we don't want
+/// to compute the track occupancy for an invalid train schedule.
+/// It's not a problem because we are just looking for the track occupancy
+/// of the operational point for the valid trains.
+pub fn extract_pathfinding_data(
+    pathfinding: &PathfindingResult,
+) -> Option<(&Vec<core_client::pathfinding::TrackRange>, &Vec<u64>)> {
+    match pathfinding {
+        PathfindingResult::Success(pathfinding_result_success) => Some((
+            &pathfinding_result_success.path.track_section_ranges,
+            &pathfinding_result_success.path_item_positions,
+        )),
+        PathfindingResult::Failure(_) => None,
+    }
+}
+
+/// Extract report train from simulation result
+/// Returns None if simulation failed
+pub fn extract_simulation_report_train(simulation: &simulation::Response) -> Option<&ReportTrain> {
+    match simulation {
+        simulation::Response::Success(SimulationResponseSuccess { final_output, .. }) => {
+            Some(&final_output.report_train)
+        }
+        _ => None,
+    }
+}
+
+/// Prepared data for computing track occupancy
+pub struct TrackOccupancyData<'a> {
+    pub path_projection: PathProjection<'a>,
+    pub path_item_positions: &'a [u64],
+    pub report_train: &'a ReportTrain,
+    pub path_item_id: Option<&'a str>,
+}
+
+/// Prepare all data needed for track occupancy computation
+/// This extracts pathfinding data, simulation data, and finds the matching path item
+pub fn prepare_track_occupancy_data<'a>(
+    operational_point_id: &str,
+    path_item_cache: &PathItemCache,
+    simulation: &'a simulation::Response,
+    pathfinding: &'a PathfindingResult,
+    train_schedule: &'a schemas::TrainSchedule,
+) -> Option<TrackOccupancyData<'a>> {
+    // Extract pathfinding data
+    let (track_ranges, path_item_positions) = extract_pathfinding_data(pathfinding)?;
+    let path_projection = PathProjection::new(track_ranges);
+
+    // Extract simulation report
+    let report_train = extract_simulation_report_train(simulation)?;
+
+    // Find matching path item
+    let path_item_id = match_path_item_id_with_operational_point(
+        path_item_cache,
+        train_schedule,
+        operational_point_id,
+    );
+
+    Some(TrackOccupancyData {
+        path_projection,
+        path_item_positions,
+        report_train,
+        path_item_id,
+    })
+}
+
+/// Find track occupancies for a train at an operational point
+/// This is a generic function used by both train_schedule and paced_train modules
+pub fn find_track_occupancy_for_operational_point(
+    operational_point_id: &str,
+    operational_point_track_offsets: &[TrackOffset],
+    path_item_cache: &PathItemCache,
```

Comment:
Taking in arguent the `PathItemCache` built somewhere else and then querying it is an open door to bugs. You feed the `PathItemCache` some OP descriptors and it queries them and indexes them on its own. However, say you ask it to retrieve an OP using its UIC, you won't be able to query it using its ID. It's a broken API, but that's where we are. So building the `PathItemCache` outside this module, but querying it here **assumes** the cache was built in a very specific way. Therefore this module logic **isn't self-contained at all**. We should either build the cache here, or fix it altogether to give it a sound API.